docs/arcadia-users-group/20230314-intro-to-python-2/lesson.md

```diff
@@ -0,0 +1,863 @@
+
+---
+### 2.3 Loops using `dict`s
+
+In addition to iterating through loops using `list` data structures, we can also use `dict`s. You can understand some of the behavior of how this iteration happens using the example below:
+
+
+```python
+favorite_vegetables = {"Marko": "pumpkin", "Charlene": "potato", "Stu": "tomato"}
+
+for entry in favorite_vegetables:
+    print(entry)
+```
+
+    Marko
+    Charlene
+    Stu
+
+
+You'll notice that iterating using a **`for`** loop across a dictionary iterates over the `key`s of that dictionary.  
+If you wanted to access both the key and the value in the dictionary, you can use the `.items()` method of a `dict` data structure.
+
+
+```python
+for villager, veg in favorite_vegetables.items():
+    print(villager + "'s favorite vegetable is", veg + ".")
+```
+
+    Marko's favorite vegetable is pumpkin.
+    Charlene's favorite vegetable is potato.
+    Stu's favorite vegetable is tomato.
+
+
+Rather than using a single variable such as `entry`, because `.items()` returns two values for every entry in the dictionary, we can unpack those paired values into two separate variables: `villager`, and `veg`.
+
+---
+### 2.4 **`while`** loops
+
+Another type of helpful loop in Python is the **`while`** loop. Unlike the **`for`** loop, which requires you to explicitly state what you're iterating over, a **`while`** loop continues as long as a conditional statement is `True`. For example:
+
+
+```python
+a = [0, 1, 2, 3, 4]
+i = 0
+
+while a[i] < 3:
+    print(a[i])
+    i += 1    # The += operator adds the value on the right and updates the original variable.
+```
+
+    0
+    1
+    2
+
+
+In the above example, there are a number of different things happening.  
+Prior to the loop, we specify a variable `i`. At the end of each entry in the loop, we increment `i` by 1 in order to step through the list `a`.
+
+How do we know when to stop? Take a look at the **`while`** statement:
+```python
+while a[i] < 3:
+```
+This means that as long as the entry at index `[i]` in list `[a]` is less than 3, we continue through the loop, printing the value of `a[i]`.
+
+As soon as we arrive at `i = 3`, at the top of the **`while`** loop, we get the value `a[i] = 3`, which is NOT less than `3`. Therefore, the **`while`** loop terminates, and the value `3` is never printed.
+
+**`while`** loops can be very powerful for a class of Python functions that rely on recursion, but you can also easily end up writing functions that "hang", or run forever without stopping.
+
+## P2. Practice
+
+Let's explore using loops.
+
+> ### Practice 3
+> Write a function called `fasta_renamer()` which takes a list of filenames and renames them with the following specifications:
+> - Converts all characters to lowercase.
+> - Converts files ending with `.fasta` to `.fa`.
+> - Optionally, adds a prefix of the user's choice to the start of every file name.
+> - Returns a list of the new file names.
```

Comment:
this might be a slightly confusing example because the files aren't written anywhere on the system, and they aren't derived from running `ls` or something on the system. i think clarifying that might be enough, or you could switch the example to sample names again, or a list of genbank or uniprot identifiers or something

Reply:
That makes sense - I updated this to mention that this just a list of file name strings, not actual files, and that we'll be learning how to open/read files in the next lesson.